### ISSUE TITLE
ofThreadChannel : destroying and resetting

### ISSUE THREAD

**Issue by elliotwoods** [openframeworks/openFrameworks#4809]
Hey!

It would be great if there's an option to reset an `ofThreadChannel` i.e. set closed = false

this would be used in cases where you are dynamically joining and setting up local threads
and you need to do the same with your thread channels

since ofThreadChannel doesn't support `=` assignment, and can't be reset, this means we have to store it in a pointer and recreate it each time

also. is there any reason why `.close()` isn't called from the destructor?

**Comment by arturoc** [comment]
yeah i've needed that sometimes too, we can add a reset method, i haven't added it yet because other implementations i've seen didn't had anything similar so i wanted to make sure that there wasn't anything that would bite us after adding it in terms of thread safety... i've used a reset method in some custom implementations that i've done for other projects and i haven't found any problem so i guess it should be ok.

also not sure about calling close on destruction, i think you usually want to have full control over the thread shutdown if you leave it to RAII it's can be problematic if you aren't very careful, for example some members in the class might have been already destroyed when the channel is closed and the thread might try to access them so it's better to explicitly close the channel and wait for the thread on the destructor of your class

**Comment by elliotwoods** [comment]
with destructor:
yes i'm aware that in many situations that's not going to work out and i presume that having destructor call close, users might not be aware that they generally need to call close themselves

if there's a way to reset the threadChannel then you wouldn't need to reinstantiate the channel in most cases so the destructor goes back to being used generally in automatic (unpredictable ways)

fingers crossed on reset!

and this is a great little class.

**Comment by elliotwoods** [comment]
also we need a size(), or something to cope with backlogs
e.g. an optional argument on send(..., size_t maximumQueueSize)

i think size() is pretty basic and can allow the user to deal with it as they like a little
clear() would be nice too. and/or `trim(size_t)`

**Comment by arturoc** [comment]
yes i didn't implement that either because i haven0t seen it in any implementation. right now this is using an std queue internally but it could use some non-locking thread safe queue and in that case most implementations won't have size. even now size is just a hint since even before than the method returned it could be a different size alredy.

about trim the way this is done in other implementations is by specifying a max size and a direction in which the channel drops new frames when it's full, that way you can implement things like drop the oldest or the newest frames.

**Comment by elliotwoods** [comment]
setting the max size sounds great to me too

the issue with size and trim being called externally is that you'd have to lock twice
perhaps 

```c++
enum ofThreadChannel::TrimStrategy {
    RemoveOld,
    RemoveNew,
    NoTrim
};

ofThreadChannel::send(const T &/&& value, TrimStrategy trim = TrimStrategy::NoTrim, size_t maxSize = 0);
```

**Comment by arturoc** [comment]
i think this is something you do on creation and perhaps don't change, it could even be a different class like ofBoundedChannel or something like that that allows to set some values and drop strategy in a setup method.

i know go has similar features, we should take a look at their api and see how it's done there but i think is just a different class. that also avoids having to do unnecessary checks in the unbounded class making it faster

**Comment by elliotwoods** [comment]
i'm not familiar with go so not sure where to look for that (had a brief tour, found channels, but the language is very different indeed!)

editing the above..

```c++
enum ofThreadChannel::TrimStrategy {
    RemoveOld,
    RemoveNew,
    NoTrim
};

ofThreadChannel::send(const T &/&& value, TrimStrategy, size_t maxSize);
```
i.e. exactly the same but without the default arguments, i.e. explicitly a separate function

i can't see any logic in having that defined in a class `ofBoundedThreadChannel` which inherits `ofThreadChannel`, since it does everything with `ofThreadChannel` does without any performance hindrence.

Anyway, the first step would be setting the `queue` and `mutex` not as private, but as `protected` which would make it possible for users to make their own `ofBoundedThreadChannel`, but I haeven't yet understood the benefit of disallowing this type of usage with `ofThreadChannel` directly

For reference, my use case is that an `ofThreadChannel` is used to communicate from a worker thread back to the main thread. If the user forgets to call `update`, then the thread channel floods, until the computer runs out of memory and becomes unstable.

Elliot

**Comment by arturoc** [comment]
actually go seems to use the same object for everything, not sure where i got the idea it was different. in any case i think setting it in a setup method rather than on each send would make more sense? it sems like having the possibility of using different strategies on each send could make the behaviour less predictable

there's also a 4th case that can be useful, that is block on send if the channel is full which allows to control the rate of the sender from the receiver

**Comment by elliotwoods** [comment]
good call!
so maybe a `setMaxBufferSize(size_t, TrimStrategy); / getMaxBufferSize();`?
And add `BlockSendWhileFull` to `TrimStrategy`?

**Comment by elliotwoods** [comment]
the advantage of having it within the `send` function, is that the existing `send` functions wouldn't check at all if there was a `TrimStrategy` or not (since it's not a state of the class but the function call)

**Comment by arturoc** [comment]
ah ok i thought you were talking about replacing the current send with the one you are proposing, having several send functions makes sense, even the strategy parameter could just be a different function. we could also have trySend to be more consistent with receive/tryReceive:

```
bool send(T&&)
bool send(T&&, int maxSize);
bool trySendDiscardLast(T&&, int maxSize);
bool trySendDiscardFirst(T&&, int maxSize);
```

or something like that. another problem with this is that the return value now means channel closed or channel full but we don't have any way to know which one it is so you can't rely on that to finish the thread which is a really nice pattern.

We could return an enum though that send returns to test if it the channel was closed or it just was full

**Comment by elliotwoods** [comment]
here's another c++ implementation (although much much longer than the one here!)
https://github.com/cameron314/readerwriterqueue/blob/master/readerwriterqueue.h

**Comment by elliotwoods** [comment]
would there be any disadvantage to setting the body of `send(const T&)` to:

```c++
	bool send(const T & value){
		T valueCopy = value;
		return this->trySend(move(value));
	}

```
?

This reduces repetition in the codebase
The only difference i can see is that you'd avoid making that copy with the existing code base iff the channel is closed.

**Comment by elliotwoods** [comment]
OK i've made a draft (without using the hint from the last comment):
https://github.com/openframeworks/openFrameworks/pull/4824

**Comment by elliotwoods** [comment]
Wait? We still don't have a `size()` on `ofThreadChannel`? (i just hacked it in on my own local copy, updating to 0.10.0 now)

It seems it was added at https://github.com/openframeworks/openFrameworks/pull/5097
But removed again?

**Comment by elliotwoods** [comment]
Interestingly, the code committed by @bakercp tries to lock a mutex in a const function (which generally disallowed as it mutates the mutex).

**Comment by bakercp** [comment]
Also, for reference, dlib has a pipe implementation that works well, but has a few differences:

http://dlib.net/dlib/pipe/pipe_kernel_1.h.html

**Comment by elliotwoods** [comment]
knock knock - can I haz size?

```c++
	/// \brief Queries size of queue.
	///
	/// This call is only an approximation, since messages come from a different
	/// thread.
	size_t size() const {
		return queue.size();
	}
```
